src/exchange/providers/retry_with_back_off_decorator.py

```diff
@@ -0,0 +1,56 @@
+import time
+from functools import wraps
+from typing import Any, Callable, Dict, Iterable, List, Optional
+
+from httpx import HTTPStatusError, Response
+
+
+def retry_with_backoff(
+        should_retry: Callable, 
+        max_retries: Optional[int] = 5, 
+        initial_wait: Optional[float] = 10, 
+        backoff_factor: Optional[float] = 1, 
+        handle_retry_exhausted: Optional[Callable] = None) ->  Callable:
+    def decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(*args: List, **kwargs: Dict) -> Any:  # noqa: ANN401
+            result = None
+            for retry in range(max_retries):
+                result = func(*args, **kwargs)
+                if not should_retry(result):
+                    return result
+                if (retry + 1) == max_retries:
+                    break
+                sleep_time = initial_wait + (backoff_factor * (2 ** retry))
+                time.sleep(sleep_time)
+            if handle_retry_exhausted:
+                handle_retry_exhausted(result, max_retries)
+            return result
+        return wrapper
+    return decorator
+
+def retry_httpx_request(
+    retry_on_status_code: Optional[Iterable[int]] = None,
+    max_retries: Optional[int] = 5,
+    initial_wait: Optional[float] = 10,
+    backoff_factor: Optional[float] = 1,
+) -> Callable:
+    if retry_on_status_code is None:
+        retry_on_status_code = set(range(400, 999))
+    def should_retry(response: Response) -> bool:
+        return response.status_code in retry_on_status_code
+
+    def handle_retry_exhausted(response: Response, max_retries: int) -> None:
+        raise HTTPStatusError(
+            f"Failed after {max_retries} retries due to flaky network or rate limiting.",
```

Comment:
I'm not sure this error message is correct - it could also be an error code we didn't anticipate that is indicative of a problem on our end (e.g. invalid API keys).

Reply:
Agree; truncating to `Failed after {max_retries} retries`. As long as we're surfacing the raw error too, this should be enough?

Reply:
fixed in 3219bc6